validator/client/src/main/java/tech/pegasys/teku/validator/client/DoppelgangerDetectionService.java

```diff
@@ -0,0 +1,181 @@
+import tech.pegasys.teku.infrastructure.async.Cancellable;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.time.TimeProvider;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.service.serviceutils.Service;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+import tech.pegasys.teku.validator.beaconnode.GenesisDataProvider;
+
+public class DoppelgangerDetectionService extends Service {
+
+  private static final Logger LOGGER = LogManager.getLogger();
+  private final Duration checkDelay;
+  private final Duration timeout;
+  private final AsyncRunner asyncRunner;
+  private final ValidatorApiChannel validatorApiChannel;
+  private final ValidatorIndexProvider validatorIndexProvider;
+  private final Spec spec;
+  private final TimeProvider timeProvider;
+  private final GenesisDataProvider genesisDataProvider;
+  private Optional<UInt64> epochAtStart = Optional.empty();
+  private volatile Cancellable doppelgangerDetectionTask;
+  private final AtomicBoolean doppelgangerCheckFinished;
+  private long startTime;
+
+  public DoppelgangerDetectionService(
+      final AsyncRunner asyncRunner,
+      final ValidatorApiChannel validatorApiChannel,
+      final ValidatorIndexProvider validatorIndexProvider,
+      final Spec spec,
+      final TimeProvider timeProvider,
+      final GenesisDataProvider genesisDataProvider,
+      final Duration checkDelay,
+      final Duration timeout) {
+    this.asyncRunner = asyncRunner;
+    this.validatorApiChannel = validatorApiChannel;
+    this.validatorIndexProvider = validatorIndexProvider;
+    this.spec = spec;
+    this.timeProvider = timeProvider;
+    this.genesisDataProvider = genesisDataProvider;
+    this.doppelgangerCheckFinished = new AtomicBoolean(false);
+    this.checkDelay = checkDelay;
+    this.timeout = timeout;
+  }
+
+  @Override
+  protected SafeFuture<?> doStart() {
+    startTime = System.nanoTime();
+    doppelgangerDetectionTask =
+        asyncRunner.runWithFixedDelay(
+            this::checkValidatorsDoppelganger,
+            checkDelay,
+            checkDelay,
+            throwable -> LOGGER.error("Error while checking validators doppelgangers.", throwable));
+    while (true) {
+      if (doppelgangerCheckFinished.get()) {
+        return SafeFuture.COMPLETE;
+      }
+    }
+  }
+
+  @Override
+  protected SafeFuture<?> doStop() {
+    LOGGER.info("No validator doppelganger detected. Stopping doppelganger detection service.");
+    return SafeFuture.fromRunnable(
+        () -> Optional.ofNullable(doppelgangerDetectionTask).ifPresent(Cancellable::cancel));
+  }
+
+  private SafeFuture<?> checkValidatorsDoppelganger() {
+    Duration duration = Duration.of(System.nanoTime() - startTime, ChronoUnit.NANOS);
+    if (duration.compareTo(timeout) > 0) {
+      LOGGER.info("Doppelganger Detection Service max allowed duration exceeded.");
```

Comment:
i think a message here is worthwhile, but we probably want to be clear about what seeing this message means from a user perspective...

Reply:
Added more details about the timeout here